src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -425,7 +425,7 @@ public static unsafe void AddScalarU(float scalar, Span<float> dst)
 
                 Vector256<float> scalarVector256 = Vector256.Create(scalar);
 
-                while (pDstCurrent + 8 <= pDstEnd)
+                while (pDstCurrent <= pDstEnd - 8)
```

Comment:
If the destination span is empty, wouldn't `pDstEnd - 8` integer underflow back to a very large number, which could lead to an AV?

Reply:
The spans passed into these functions are guaranteed not to be empty in the calling code.

https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

Reply:
This is also guaranteed in release builds? The contracts routines look like they're surrounded by `[Conditional("DEBUG")]`.

Reply:
Yes, it is an invariant that these methods are not called with empty spans. Even higher in the callstack (in many places) are the real checks. For example:

https://github.com/dotnet/machinelearning/blob/d38a35edda4e3b3bb9e74fd78f57e7a49a906186/src/Microsoft.ML.Transforms/GcnTransform.cs#L479-L487

Here, when `count == 0`, the code doesn't even get into `CpuMathUtils.Add` below - it exits early.

On the "other side" of these CpuMathUtils (when are are compiling for `netstandard` where C# intrinsics don't exist), we make similar assumptions:

https://github.com/dotnet/machinelearning/blob/a570da14a41f2870eb8f61d84496a58422398253/src/Microsoft.ML.CpuMath/CpuMathUtils.netstandard.cs#L95-L101

`MemoryMarshal.GetReference` is going to fail if you pass in an empty span.

Reply:
Also note - the CpuMathUtils methods are not public.